scio-core/src/main/scala/com/spotify/scio/transforms/ScalaAsyncLookupDoFn.scala

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.transforms
+
+import java.util.UUID
+import java.util.concurrent.Semaphore
+import java.util.function.{Function => JFunction}
+
+import com.google.common.cache.Cache
+import com.google.common.collect.{Maps, Queues}
+import org.apache.beam.sdk.values.KV
+import org.apache.beam.sdk.transforms.DoFn
+import org.apache.beam.sdk.transforms.DoFn.{FinishBundle, ProcessElement, Setup, StartBundle}
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow
+import org.joda.time.Instant
+import com.google.common.base.Preconditions
+import com.google.common.util.concurrent.{FutureCallback, Futures, ListenableFuture, MoreExecutors}
+import com.spotify.scio.transforms.ScalaAsyncLookupDoFn.Caches
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Await, Future, Promise}
+import scala.concurrent.duration.Duration
+import scala.util.{Failure, Success, Try}
+import scala.language.implicitConversions
+import scala.util.control.NonFatal
+
+object ScalaAsyncLookupDoFn {
+  val Client = Maps.newConcurrentMap[UUID, Any]
+  val Caches = Maps.newConcurrentMap[UUID, Cache[_, _]]
+
+  implicit def toJavaFunction[A, B](f: Function1[A, B]): JFunction[A, B] =
```

Comment:
https://github.com/spotify/scio/pull/1824/files#discussion_r274093420

You said we want a Scala AsyncDoFn to avoid converting between Java and Scala futures, and yet we're doing here again?

Also avoid implicit conversion unless absolute necessary.